src/main/scala/io/iohk/ethereum/domain/BlockchainReader.scala

```diff
@@ -113,6 +111,38 @@ class BlockchainReader(
   def genesisBlock: Block =
     getBlockByNumber(0).get
 
+  /** Returns a block inside this branch based on its number */
+  def getBlockByNumber(branch: Branch, number: BigInt): Option[Block] = branch match {
+    case BestBranch(_, tipBlockNumber) =>
+      if (tipBlockNumber >= number && number >= 0) {
```

Comment:
Could be nice as a guard on the `case`.

Reply:
I rewrote it to have the guard, however I refrained from having a `case _ =>` below because I would like to get the exhaustive check to fail clearly in case we modify the trait later on.